docs/tutorials/PyTorch.py

```diff
@@ -0,0 +1,322 @@
+a_tensor = torch.tensor(an_array)  # creates tensor of same size as an_array
+
+print(a_tensor)
+# -
+
+# Tensors have many parallels to numpy arrays. There are several [operations](https://pytorch.org/docs/stable/torch.html) that we can perform on tensors that we'd usually perform on numpy arrays. For example, we can multiply two numpy arrays and compare this with the multiplication of the corresponding PyTorch tensors:
+
+# +
+another_array = np.array([[5, 6, 7], [8, 9, 0]])  # create 2x3 array
+another_tensor = torch.tensor(
+    another_array
+)  # create another tensor of same size as above array
+
+# numpy array multiplication
+prod_array = np.matmul(an_array, another_array)
+
+# torch tensor multiplication
+prod_tensor = torch.matmul(a_tensor, another_tensor)
+
+print(f"Numpy array multiplication result: {prod_array}")
+print(f"Torch tensor multiplication result: {prod_tensor}")
+# -
+
+# ## Calculating Gradients
+
+# PyTorch provides a key functionality---the ability to calculate the gradients on tensors. Let's start by creating a tensor with a single value. Here we are setting <code> requires_grad = True </code>, we'll see why this is important in a moment.
+
+x = torch.tensor(3.0, requires_grad=True)
+x
+
+# Let's define some variable y in terms of x:
+
+y = x ** 2
+
+#  We see that the value of ``y`` is as we expect---nothing too strange here.
+
+print(f"y: {y}")
+print(f"x: {x}")
+
+# But what if we wanted to calculate the gradient of ``y`` with respect to ``x``? Using calculus, we find that the answer is dy/dx = 2*x. The derivative evaluated at ``x=3`` is ``6``.
+#
+# We can see if PyTorch gets the same answer as us if we do:
+
+y.backward()  # populates gradient (.grad) attributes of y with respect to all of its independent variables
+x.grad  # returns the grad attribute (the gradient) of y with respect to x
+
+# PyTorch uses the concept of automatic differentiation to calculate the derivative. Instead of computing the derivative as we would by hand, the program is using a computational graph.
+
+# ## Optimizing a Function with Gradient Descent
+#
+# If we were on the side of a hill in the dark and we wanted to get down to the bottom of a valley, how would we do it?
+#
+# We wouldn't be able to see all the way to the bottom of the valley, but we could feel which way is down based on where we are standing. We would take steps in the downward direction and we'd know when to stop when the ground felt flat.
+#
+# One other thing we'd have to consider is our step size. If we take very small steps in the direction of the descent, it will take us a longer time than if we take larger steps. However, if we take super long steps, we might completely miss the flat part of the valley, and start ascending the other side of the valley.
+
+# We can look at the gradient descent from a more mathematical lense by looking at the graph z = t ** 2:
+
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+# We start on the left hill at the point (-4, 16):
```

Comment:
Have you considered starting at a random point? I think that might make it easier for you to write this in a general way rather than hard-coding so many values.

Reply:
I think it's OK to start at a pre-chosen point, since it looks visually pretty nice. If you were to choose a random point, I think you'd want to first set the random seed to a constant, such that the tutorial is repeatable each time. And then if you're doing this tuning, you might as well just choose a point in the first place.